File: gcj/2009.1A/A.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

#define MAX_N 200000000

static bool happiness_cache[11][MAX_N], cached[11][MAX_N];

static inline int sum_of_square_digits(int n, int b) {
	int m = 0;
	while (n) {
		m += (n%b) * (n%b);
		n /= b;
	}
	return m;
}

static inline bool calculate_happiness(int n, int b) {
	set<int> s;
	while (n != 1 && s.find(n) == s.end()) {
		s.insert(n);
		n = sum_of_square_digits(n,b);
	}
	return n == 1;
}

static inline void init() {
	for (int b = 2; b < 11; b++) {
		happiness_cache[b][1] = true;
		cached[b][n] = true;
		for (int n = 2; n < 1000; n++) {
			happiness_cache[b][n] = calculate_happiness(n);
			cached[b][n] = true;
		}
	}
}

static inline bool happy(int n, int b) {
	if (!cached[b][n]) {
		happiness_cache[b][n] = happy(sum_of_square_digits(n,b),b);
		cached[b][n] = true;
	}
	return happiness_cache[b][n];
}

int main() {
	init();
	
}

File: gcj/2009.1A/A.py
N_LIMIT = 100000
B_LIMIT = 11
MEMO = [ None ] * B_LIMIT

def init():
	for b in range(2,11):
		h = [None] * N_LIMIT
		h[1] = True
		for n in range(2,b ** 3):
			h[n] = calculate_happiness(n,b)
		
		for n in range(b,N_LIMIT):
			h[n] = h[sum_of_square_digits(n,b)]
		
		MEMO[b] = h

def sum_of_square_digits(n,b):
	m = 0
	while n:
		m += (n % b) ** 2
		n //= b
	return m

def calculate_happiness(n,b):
	s = set()
	while n != 1 and n not in s:
		s.add(n)
		n = sum_of_square_digits(n,b)
	return n == 1

def faster_calculate_happiness(n,b):
	while n >= N_LIMIT:
		n = sum_of_square_digits(n,b)
	return MEMO[b][n]

# for n in range(1000):
# 	print("%-4s: %s"% (n,calculate_happiness(n,10)))

print([n for n in range(1000) if calculate_happiness(n,10)])
